Mathlib/Algebra/Lie/Weights/Chain.lean

```diff
@@ -217,4 +217,124 @@ end IsCartanSubalgebra
+@[simp] lemma chainBotCoeff_neg : chainBotCoeff (-α) β = chainTopCoeff α β := by
+  rw [← chainTopCoeff_neg, neg_neg]
+
+@[simp] lemma chainTopCoeff_zero : chainTopCoeff 0 β = 0 := dif_pos rfl
+@[simp] lemma chainBotCoeff_zero : chainBotCoeff 0 β = 0 := dif_pos neg_zero
+
+lemma chainTopCoeff_add_one :
+    letI := Classical.propDecidable
+    chainTopCoeff α β + 1 =
+      Nat.find (eventually_weightSpace_smul_add_eq_bot M α β hα).exists := by
+  classical
+  rw [chainTopCoeff, dif_neg hα]
+  apply Nat.succ_pred_eq_of_pos
+  rw [zero_lt_iff]
+  intro e
+  have : weightSpace M (0 • α + β : L → R) = ⊥ := by
+    rw [← e]
+    exact Nat.find_spec (eventually_weightSpace_smul_add_eq_bot M α β hα).exists
+  exact β.weightSpace_ne_bot _ (by simpa only [zero_smul, zero_add] using this)
+
+lemma weightSpace_chainTopCoeff_add_one_nsmul_add :
+    weightSpace M ((chainTopCoeff α β + 1) • α + β : L → R) = ⊥ := by
+  classical
+  rw [chainTopCoeff_add_one _ _ hα]
+  exact Nat.find_spec (eventually_weightSpace_smul_add_eq_bot M α β hα).exists
+
+lemma weightSpace_chainTopCoeff_add_one_zsmul_add :
+    weightSpace M ((chainTopCoeff α β + 1 : ℤ) • α + β : L → R) = ⊥ := by
+  rw [← weightSpace_chainTopCoeff_add_one_nsmul_add α β hα, nsmul_eq_smul_cast ℤ,
+    Nat.cast_add, Nat.cast_one]
+
+lemma weightSpace_nsmul_add_ne_bot_of_le {n} (hn : n ≤ chainTopCoeff α β) :
+    weightSpace M (n • α + β : L → R) ≠ ⊥ := by
+  by_cases hα : α = 0
+  · rw [hα, smul_zero, zero_add]; exact β.weightSpace_ne_bot
+  classical
+  rw [← Nat.lt_succ, Nat.succ_eq_add_one, chainTopCoeff_add_one _ _ hα] at hn
+  exact Nat.find_min (eventually_weightSpace_smul_add_eq_bot M α β hα).exists hn
+
+lemma weightSpace_chainBotCoeff_sub_one_zsmul_sub :
+    weightSpace M ((-chainBotCoeff α β - 1 : ℤ) • α + β : L → R) = ⊥ := by
+  rw [sub_eq_add_neg, ← neg_add, neg_smul, ← smul_neg, chainBotCoeff,
+    weightSpace_chainTopCoeff_add_one_zsmul_add _ _ (by simpa using hα)]
+
+lemma weightSpace_zsmul_add_ne_bot {n : ℤ}
+    (hn : -chainBotCoeff α β ≤ n) (hn' : n ≤ chainTopCoeff α β) :
+      weightSpace M (n • α + β : L → R) ≠ ⊥ := by
+  rcases n with (n | n)
+  · simp only [Int.ofNat_eq_coe, Nat.cast_le, ← nsmul_eq_smul_cast] at hn' ⊢
+    exact weightSpace_nsmul_add_ne_bot_of_le α β hn'
+  · simp only [Int.negSucc_eq, ← Nat.cast_succ, neg_le_neg_iff, Nat.cast_le] at hn ⊢
+    rw [neg_smul, ← smul_neg, ← nsmul_eq_smul_cast]
+    exact weightSpace_nsmul_add_ne_bot_of_le (-α) β hn
+
+lemma weightSpace_neg_zsmul_add_ne_bot {n : ℕ} (hn : n ≤ chainBotCoeff α β) :
+    weightSpace M ((-n : ℤ) • α + β : L → R) ≠ ⊥ := by
+  apply weightSpace_zsmul_add_ne_bot α β <;> omega
+
+/-- This is the first weight in an `α`-chain through `β`. -/
+noncomputable
+def chainTop (α : L → R) (β : Weight R L M) : Weight R L M :=
+  ⟨chainTopCoeff α β • α + β, weightSpace_nsmul_add_ne_bot_of_le α β le_rfl⟩
+
+/-- This is the first weight in an `α`-chain through `β`. -/
```

Comment:
These two docstrings are identical. I guess you want `s/first/last/` in the first docstring?

Reply:
Also, I would start the docstrings with `The .. weight`.
```suggestion
/-- The last weight in an `α`-chain through `β`. -/
noncomputable
def chainTop (α : L → R) (β : Weight R L M) : Weight R L M :=
  ⟨chainTopCoeff α β • α + β, weightSpace_nsmul_add_ne_bot_of_le α β le_rfl⟩

/-- The first weight in an `α`-chain through `β`. -/
```